### PR TITLE
usb:  fix coverity issues, fix ecm class, alternative to #14977

### PR DESCRIPTION
**Pull request by jfischer-no**
usb:  fix coverity issues, fix ecm class, alternative to #14977

I have looked at all the drivers again, and everyone except usb_dc_kinetis checks
in his own way MPS (but not all if it fits the endpoint type), so I have added the checks only to usb_dc_kinetis.

cherry-picked from #14977, #14924

### CODE REVIEW

File: include/usb/usb_device.h [hunk 1]
```diff
@@ -74,6 +74,14 @@ extern "C" {
 
 #define MAX_PACKET_SIZE0    64        /**< maximum packet size for EP 0 */
 
+#define USB_MPS_MAX_CTRL	64        /**< maximum packet size for EP 0 */
```
Comment (finikorg) [inline include/usb/usb_device.h]: I think that more readable is `USB_MAX_CTRL_MPS`

File: drivers/usb/device/usb_dc_kinetis.c [hunk 2]
```diff
@@ -309,8 +330,7 @@ int usb_dc_ep_configure(const struct usb_dc_ep_cfg_data * const cfg)
 	struct k_mem_block *block;
 	struct usb_ep_ctrl_data *ep_ctrl = &dev_data.ep_ctrl[ep_idx];
 
-	if (ep_idx > (NUM_OF_EP_MAX - 1)) {
-		LOG_ERR("Wrong endpoint index/address");
+	if (usb_dc_ep_check_cap(cfg)) {
```
Comment (finikorg) [inline drivers/usb/device/usb_dc_kinetis.c]: Do you want to move `usb_dc_ep_check_cap()` inside `usb_dc_ep_configure()`? Then it needs to be done for every dc.

**Review by finikorg — CHANGES_REQUESTED** [review]

File: include/usb/usb_device.h (hunk 1 above)
Comment (jfischer-phytec-iot) [reply include/usb/usb_device.h]: okay

File: drivers/usb/device/usb_dc_kinetis.c (hunk 2 above)
Comment (jfischer-phytec-iot) [reply drivers/usb/device/usb_dc_kinetis.c]: I will, but probably not for 1.14

File: drivers/usb/device/usb_dc_kinetis.c [hunk 3]
```diff
@@ -276,14 +276,35 @@ int usb_dc_ep_check_cap(const struct usb_dc_ep_cfg_data * const cfg)
...
-	}
-
 	if (ep_idx > (NUM_OF_EP_MAX - 1)) {
 		LOG_ERR("endpoint index/address out of range");
-		return -1;
+		return -EINVAL;
+	}
+
+	switch (cfg->ep_type) {
+	case USB_DC_EP_CONTROL:
```
Comment (finikorg) [inline drivers/usb/device/usb_dc_kinetis.c]: For the control EP we have now `USB_MAX_CTRL_MPS`, right?

File: drivers/usb/device/usb_dc_kinetis.c [hunk 4]
```diff
@@ -276,14 +276,35 @@ int usb_dc_ep_check_cap(const struct usb_dc_ep_cfg_data * const cfg)
...
+		if (cfg->ep_mps > USB_MAX_FS_ISO_MPS) {
+			return -EINVAL;
+		}
+		break;
+	default:
+		LOG_ERR("Unknown endpoint type!");
+		return -EINVAL;
+	}
+
+	if (cfg->ep_type == USB_DC_EP_CONTROL) {
```
Comment (finikorg) [inline drivers/usb/device/usb_dc_kinetis.c]: this is not needed, can be handled in the switch above

File: drivers/usb/device/usb_dc_kinetis.c (hunk 2 above)
Comment (finikorg) [reply drivers/usb/device/usb_dc_kinetis.c]: OK, good enough

File: subsys/usb/usb_device.c [hunk 5]
```diff
@@ -549,18 +549,29 @@ static bool usb_set_configuration(u8_t config_index, u8_t alt_setting)
 static bool usb_set_interface(u8_t iface, u8_t alt_setting)
 {
 	const u8_t *p = usb_dev.descriptors;
-	u8_t cur_iface = 0xFF;
+	const u8_t *if_desc = NULL;
 	u8_t cur_alt_setting = 0xFF;
-	struct usb_dc_ep_cfg_data ep_cfg;
+	u8_t cur_iface = 0xFF;
```
Comment (jukkar) [inline subsys/usb/usb_device.c]: Is this the network interface index value or some other value that is specific to USB? The reason I am asking, is that currently network interface index numbering starts from 1, and 0 value can be used to mean unknown or unset value.

Actually it is a bit confusing to read this commit as typically iface is a pointer to struct net_if and if_index, if_idx or similar is the network interface index value. Now if_desc is a pointer and iface is index, so reviewing this needs some more effort.

Comment (finikorg) [reply subsys/usb/usb_device.c]: This is not network interface, this is USB interface number.

Comment (jfischer-phytec-iot) [reply subsys/usb/usb_device.c]: @jukkar It is specific to USB, means usb device interface (usb_device contains usb device code only)

**Review by finikorg — APPROVED** [review]
LGTM

**Review by nashif — APPROVED** [review]

**Pull request merged**